File: app/src/main/java/io/ipoli/android/app/utils/DateUtils.java
package io.ipoli.android.app.utils;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import org.threeten.bp.DayOfWeek;
import org.threeten.bp.Instant;
import org.threeten.bp.LocalDate;
import org.threeten.bp.Month;
import org.threeten.bp.ZoneId;
import org.threeten.bp.ZonedDateTime;
import org.threeten.bp.format.TextStyle;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import static org.threeten.bp.temporal.TemporalAdjusters.firstDayOfMonth;
import static org.threeten.bp.temporal.TemporalAdjusters.lastDayOfMonth;

public class DateUtils {

    public static final ZoneId ZONE_UTC = ZoneId.of("UTC");

    public static LocalDate fromMillis(long dateMillis) {
        return Instant.ofEpochMilli(dateMillis).atZone(DateUtils.ZONE_UTC).toLocalDate();
    }

    public static LocalDate fromMillis(long dateMillis, ZoneId zoneId) {
        return Instant.ofEpochMilli(dateMillis).atZone(zoneId).toLocalDate();
    }

    public static String getShortName(Month month) {
        return month.getDisplayName(TextStyle.SHORT, Locale.getDefault());
    }

    public static String getMonthShortName(LocalDate date) {
        return getShortName(date.getMonth());
    }

    public static Date now() {
        return Calendar.getInstance().getTime();
    }

    public static boolean isToday(LocalDate date) {
        return LocalDate.now().isEqual(date);
    }

    public static boolean isTomorrow(LocalDate date) {
        return LocalDate.now().plusDays(1).isEqual(date);
    }

    public static Date getTomorrow() {
        Calendar c = Calendar.getInstance();
        c.add(Calendar.DAY_OF_YEAR, 1);
        return c.getTime();
    }

    public static Date nowUTC() {
        return new Date(System.currentTimeMillis());
    }

    public static Date toStartOfDayUTC(LocalDate localDate) {
        return fromZonedDateTime(localDate.atStartOfDay(ZONE_UTC));
    }

    public static Date toStartOfDay(LocalDate localDate) {
        return fromZonedDateTime(localDate.atStartOfDay(ZoneId.systemDefault()));
    }

    public static long toMillis(LocalDate localDate) {
        return toStartOfDayUTC(localDate).getTime();
    }

    private static Date fromZonedDateTime(ZonedDateTime dateTime) {
        return new Date(dateTime.toInstant().toEpochMilli());
    }

    public static boolean isTodayUTC(LocalDate localDate) {
        return localDate.isEqual(toStartOfDayUTCLocalDate(LocalDate.now()));
    }

    public static boolean isTomorrowUTC(LocalDate localDate) {
        return localDate.isEqual(toStartOfDayUTCLocalDate(LocalDate.now().plusDays(1)));
    }

    public static LocalDate toStartOfDayUTCLocalDate(LocalDate localDate) {
        return localDate.atStartOfDay(ZONE_UTC).toLocalDate();
    }

    public static boolean isBetween(LocalDate date, LocalDate start, LocalDate end) {
        if (date == null || start == null || end == null) {
            return false;
        }
        return !date.isBefore(start) && !date.isAfter(end);
    }

    @NonNull
    public static List<Pair<LocalDate, LocalDate>> getBoundsFor4MonthsInThePast(LocalDate currentDate) {
        LocalDate monthStart = currentDate.minusMonths(3).with(firstDayOfMonth());
        LocalDate monthEnd = monthStart.with(lastDayOfMonth());

        List<Pair<LocalDate, LocalDate>> monthBounds = new ArrayList<>();
        monthBounds.add(new Pair<>(monthStart, monthEnd));
        for (int i = 0; i < 3; i++) {
            monthStart = monthStart.plusMonths(1);
            monthEnd = monthStart.with(lastDayOfMonth());
            monthBounds.add(new Pair<>(monthStart, monthEnd));
        }
        return monthBounds;
    }

    @NonNull
    public static List<Pair<LocalDate, LocalDate>> getBoundsFor4WeeksInThePast(LocalDate currentDate) {
        LocalDate weekStart = currentDate.minusWeeks(3).with(DayOfWeek.MONDAY);
        LocalDate weekEnd = weekStart.with(DayOfWeek.SUNDAY);

        List<Pair<LocalDate, LocalDate>> weekBounds = new ArrayList<>();
        weekBounds.add(new Pair<>(weekStart, weekEnd));
        for (int i = 0; i < 3; i++) {
            weekStart = weekStart.plusWeeks(1);
            weekEnd = weekStart.with(DayOfWeek.SUNDAY);
            weekBounds.add(new Pair<>(weekStart, weekEnd));
        }
        return weekBounds;
    }

    public static boolean isYesterday(LocalDate date) {
        return LocalDate.now().minusDays(1).isEqual(date);
    }

    public static String getDayNumberSuffix(int day) {
        if (day >= 11 && day <= 13) {
            return "th";
        }
        switch (day % 10) {
            case 1:
                return "st";
            case 2:
                return "nd";
            case 3:
                return "rd";
            default:
                return "th";
        }
    }

    public static LocalDate fromUserZoneToLocalDate(Date date) {
        return Instant.ofEpochMilli(date.getTime()).atZone(ZoneId.systemDefault()).toLocalDate();
    }
}

File: app/src/main/java/mypoli/android/common/datetime/DateUtils.kt
package mypoli.android.common.datetime

import android.util.Pair
import org.threeten.bp.*
import org.threeten.bp.format.TextStyle
import org.threeten.bp.temporal.TemporalAdjusters.firstDayOfMonth
import org.threeten.bp.temporal.TemporalAdjusters.lastDayOfMonth
import org.threeten.bp.temporal.WeekFields
import java.util.*

/**
 * Created by Venelin Valkov <venelin@mypoli.fun>
 * on 8/20/17.
 */
object DateUtils {

    val ZONE_UTC = ZoneId.of("UTC")

    /**
     * @param dateMillis in UTC timezone
     * @return LocalDate
     */
    fun fromMillis(dateMillis: Long): LocalDate {
        return Instant.ofEpochMilli(dateMillis).atZone(DateUtils.ZONE_UTC).toLocalDate()
    }

    fun fromMillis(dateMillis: Long, zoneId: ZoneId): LocalDate {
        return Instant.ofEpochMilli(dateMillis).atZone(zoneId).toLocalDate()
    }

    fun getShortName(month: Month): String {
        return month.getDisplayName(TextStyle.SHORT, Locale.getDefault())
    }

    fun getMonthShortName(date: LocalDate): String {
        return getShortName(date.month)
    }

    fun now(): Date {
        return Calendar.getInstance().time
    }

    fun isToday(date: LocalDate): Boolean {
        return LocalDate.now().isEqual(date)
    }

    fun isTomorrow(date: LocalDate): Boolean {
        return LocalDate.now().plusDays(1).isEqual(date)
    }

    fun nowUTC(): Date {
        return Date(System.currentTimeMillis())
    }

    fun toStartOfDayUTC(localDate: LocalDate): Date {
        return fromZonedDateTime(localDate.atStartOfDay(ZONE_UTC))
    }

    fun toStartOfDay(localDate: LocalDate): Date {
        return fromZonedDateTime(localDate.atStartOfDay(ZoneId.systemDefault()))
    }

    /**
     * @param localDate
     * @return timestamp in UTC
     */
    fun toMillis(localDate: LocalDate): Long {
        return toStartOfDayUTC(localDate).time
    }

    private fun fromZonedDateTime(dateTime: ZonedDateTime): Date {
        return Date(dateTime.toInstant().toEpochMilli())
    }

    fun isTodayUTC(localDate: LocalDate): Boolean {
        return localDate.isEqual(toStartOfDayUTCLocalDate(LocalDate.now()))
    }

    fun isTomorrowUTC(localDate: LocalDate): Boolean {
        return localDate.isEqual(toStartOfDayUTCLocalDate(LocalDate.now().plusDays(1)))
    }

    fun toStartOfDayUTCLocalDate(localDate: LocalDate): LocalDate {
        return localDate.atStartOfDay(ZONE_UTC).toLocalDate()
    }

    fun isBetween(date: LocalDate?, start: LocalDate?, end: LocalDate?): Boolean {
        return if (date == null || start == null || end == null) {
            false
        } else !date.isBefore(start) && !date.isAfter(end)
    }

    fun getBoundsFor4MonthsInThePast(currentDate: LocalDate): List<Pair<LocalDate, LocalDate>> {
        var monthStart = currentDate.minusMonths(3).with(firstDayOfMonth())
        var monthEnd = monthStart.with(lastDayOfMonth())

        val monthBounds = ArrayList<Pair<LocalDate, LocalDate>>()
        monthBounds.add(Pair(monthStart, monthEnd))
        for (i in 0..2) {
            monthStart = monthStart.plusMonths(1)
            monthEnd = monthStart.with(lastDayOfMonth())
            monthBounds.add(Pair(monthStart, monthEnd))
        }
        return monthBounds
    }

    fun getBoundsFor4WeeksInThePast(currentDate: LocalDate): List<Pair<LocalDate, LocalDate>> {
        var weekStart = currentDate.minusWeeks(3).with(DayOfWeek.MONDAY)
        var weekEnd = weekStart.with(DayOfWeek.SUNDAY)

        val weekBounds = ArrayList<Pair<LocalDate, LocalDate>>()
        weekBounds.add(Pair(weekStart, weekEnd))
        for (i in 0..2) {
            weekStart = weekStart.plusWeeks(1)
            weekEnd = weekStart.with(DayOfWeek.SUNDAY)
            weekBounds.add(Pair(weekStart, weekEnd))
        }
        return weekBounds
    }

    fun isYesterday(date: LocalDate): Boolean {
        return LocalDate.now().minusDays(1).isEqual(date)
    }

    fun getDayNumberSuffix(day: Int): String {
        if (day >= 11 && day <= 13) {
            return "th"
        }
        when (day % 10) {
            1 -> return "st"
            2 -> return "nd"
            3 -> return "rd"
            else -> return "th"
        }
    }

//    fun dayOfWeekText(val dayOfWeek : DayOfWeek, style: TextStyle): String {
//        return (0 until DAYS_IN_A_WEEK).map {
//            val dayOfWeek = DateUtils.firstDayOfWeek.plus(it.toLong())
//
//            dayOfWeek.getDisplayName(
//                style,
//                Locale.getDefault()
//            )
//        }
//    }

    const val DAYS_IN_A_WEEK = 7

    fun daysOfWeekText(style: TextStyle): List<String> {
        return (0 until DAYS_IN_A_WEEK).map {
            val dayOfWeek = DateUtils.firstDayOfWeek.plus(it.toLong())
            dayOfWeekText(dayOfWeek, style)
        }
    }

    fun dayOfWeekText(
        dayOfWeek: DayOfWeek,
        style: TextStyle
    ): String {
        return dayOfWeek.getDisplayName(
            style,
            Locale.getDefault()
        )
    }

    val localeDaysOfWeek: List<DayOfWeek>
        get() = (0 until DAYS_IN_A_WEEK).map {
            DateUtils.firstDayOfWeek.plus(it.toLong())
        }

    val firstDayOfWeek: DayOfWeek
        get() = WeekFields.of(Locale.getDefault()).firstDayOfWeek

    val lastDayOfWeek: DayOfWeek
        get() = firstDayOfWeek.plus(6)

    val today: LocalDate
        get() = LocalDate.now()

    fun max(date1: LocalDate, date2: LocalDate): LocalDate {
        if (date1.isAfter(date2)) {
            return date1
        }

        return date2
    }

    fun min(date1: LocalDate, date2: LocalDate): LocalDate {
        if (date1.isBefore(date2)) {
            return date1
        }

        return date2
    }
}
